src/process_collector.rs

```diff
@@ -0,0 +1,358 @@
+            self.start_time.set(start_time);
+        }
+
+        // cpu
+        let cpu_total_mfs = {
+            let cpu_total = self.cpu_total.lock().unwrap();
+            if let Some(ref stat) = pid_stat {
+                let total = (stat.utime + stat.stime) as f64 / *CLK_TCK;
+                let past = cpu_total.get();
+                let delta = total - past;
+                if delta > 0.0 {
+                    cpu_total.inc_by(delta).unwrap();
+                }
+            }
+
+            cpu_total.collect()
+        };
+
+        // proc_start_time
+        if let (Some(ref stat), Some(boot_time)) = (pid_stat, *BOOT_TIME) {
+            self.start_time.set(stat.start_time as f64 / *CLK_TCK + boot_time);
+        }
+
+        // collect MetricFamilys.
+        let mut mfs = Vec::new();
+        mfs.extend(cpu_total_mfs);
+        mfs.extend(self.open_fds.collect());
+        mfs.extend(self.max_fds.collect());
+        mfs.extend(self.vsize.collect());
+        mfs.extend(self.rss.collect());
+        mfs.extend(self.start_time.collect());
+        mfs
+    }
+}
+
+impl Default for ProcessCollector {
+    /// Returns a `ProcessCollector` of the calling process.
+    fn default() -> Self {
+        let pid = unsafe { libc::getpid() };
+        ProcessCollector::new(pid, "")
+    }
+}
+
+fn open_fds(pid: pid_t) -> Result<usize> {
+    let path = format!("/proc/{}/fd", pid);
+    try!(fs::read_dir(path)).fold(Ok(0), |acc, i| {
+        let mut acc = try!(acc);
+        let ty = try!(try!(i).file_type());
+        if !ty.is_dir() {
+            acc += 1;
+        }
+
+        Ok(acc)
+    })
+}
+
+// `find_statistic` match lines in pattern pat, it takes the first matching line,
+// and parse the first number literal in the matching line.
+//
+// Example:
+//  * all:
+// ```
+// ctxt 789298306
+// btime 1477460662
+// processes 302136
+// procs_running 1
+// procs_blocked 0
+// ```
+//  * pat: "btime"
+//
+// then it returns `Ok(1477460662.0)`
+//
+// pub for tests.
+pub fn find_statistic(all: &str, pat: &str) -> Result<f64> {
+    let literal = all.lines()
+        .find(|line| line.contains(pat))
+        .and_then(|line| {
+            (&line[pat.len()..line.len()])
+                .split_whitespace()
+                .find(|s| !s.is_empty())
+        });
+    match literal {
+        Some(literal) => {
+            literal.parse().map_err(|e| Error::Msg(format!("read statistic {} failed: {}", pat, e)))
+        }
+        None => Err(Error::Msg(format!("read statistic {} failed", pat))),
+    }
+
+}
+
+// See more `man 5 proc`, `/proc/[pid]/limits`
+const MAX_FD_PATTERN: &'static str = "Max open files";
+
+fn max_fds(pid: pid_t) -> Result<f64> {
+    let path = format!("/proc/{}/limits", pid);
+    let mut buffer = String::new();
+    try!(fs::File::open(path).and_then(|mut f| f.read_to_string(&mut buffer)));
+    find_statistic(&buffer, MAX_FD_PATTERN)
```

Comment:
ProcessCollector needs to get other process's limits.

`getrlimit` only gets self limits, however we can use [`prlimit(2)`](https://linux.die.net/man/2/prlimit) instead.

> The Linux-specific prlimit() system call combines and extends the functionality of setrlimit() and getrlimit().  It can be used to both set and get the resource limits of an **arbitrary** process.

Oh, another difference is `getrlimit()` is faster ;-)